modules/tinymce/src/plugins/imagetools/main/ts/core/Actions.ts

```diff
@@ -20,43 +21,52 @@ import * as ImageSize from './ImageSize';
 
 let count = 0;
 
-const getFigureImg = (elem) => SelectorFind.child(SugarElement.fromDom(elem), 'img');
+const enum FileExtractType {
+  Name = 2,
+  NameExt = 1
+}
 
-const isFigure = (editor: Editor, elem) => editor.dom.is(elem, 'figure');
+const getFigureImg = (elem: HTMLElement) => SelectorFind.child<HTMLImageElement>(SugarElement.fromDom(elem), 'img');
 
-const getEditableImage = (editor: Editor, elem) => {
-  const isImage = (imgNode) => editor.dom.is(imgNode, 'img:not([data-mce-object],[data-mce-placeholder])');
-  const isEditable = (imgNode) => isImage(imgNode) && (isLocalImage(editor, imgNode) || isCorsImage(editor, imgNode) || Settings.getProxyUrl(editor));
+const isFigure = (editor: Editor, elem: Node): elem is HTMLElement => editor.dom.is(elem, 'figure');
 
-  if (isFigure(editor, elem)) {
-    const imgOpt = getFigureImg(elem);
-    return imgOpt.map((img) => isEditable(img.dom) ? Optional.some(img.dom) : Optional.none());
+const isImage = (editor: Editor, imgNode: Node): imgNode is HTMLImageElement => editor.dom.is(imgNode, 'img:not([data-mce-object],[data-mce-placeholder])');
+
+const getEditableImage = (editor: Editor, node: Node): Optional<HTMLImageElement> => {
+  const isEditable = (imgNode: Node): imgNode is HTMLImageElement =>
+    isImage(editor, imgNode) && (isLocalImage(editor, imgNode) || isCorsImage(editor, imgNode) || Type.isNonNullable(Settings.getProxyUrl(editor)));
+
+  if (isFigure(editor, node)) {
+    return getFigureImg(node).bind((img) => {
+      return isEditable(img.dom) ? Optional.some(img.dom) : Optional.none();
+    });
+  } else {
+    return isEditable(node) ? Optional.some(node) : Optional.none();
   }
-  return isEditable(elem) ? Optional.some(elem) : Optional.none();
 };
 
-const displayError = function (editor: Editor, error) {
+const displayError = function (editor: Editor, error: string) {
   editor.notificationManager.open({
     text: error,
     type: 'error'
   });
 };
 
-const getSelectedImage = (editor: Editor): Optional<SugarElement> => {
+const getSelectedImage = (editor: Editor): Optional<SugarElement<HTMLImageElement>> => {
```

Comment:
This function was rather flawed and adding the types showed that, as everywhere this was being used it assumed the returned element was an image. However, since this it was just `editor.selection.getNode()` if could have been literally anything. So this largely copies the functionality from the image plugin lookup.